core/templates/dev/head/components/question-directives/question-player/question-player.directive.ts

```diff
@@ -282,17 +294,53 @@ oppia.directive('questionPlayer', [
               }
               for (var i = 0; i < questionData.linkedSkillIds.length; i++) {
                 var skillId = questionData.linkedSkillIds[i];
-                if (!(skillId in scorePerSkillMapping)) {
+                if (!(skillId in ctrl.scorePerSkillMapping)) {
                   continue;
                 }
-                scorePerSkillMapping[skillId].score += questionScore;
+                ctrl.scorePerSkillMapping[skillId].score += questionScore;
+                ctrl.scorePerSkillMapping[skillId].total += 1.0;
               }
             }
             ctrl.totalScore = Math.round(
               ctrl.totalScore * 100 / totalQuestions);
             $scope.resultsLoaded = true;
           };
 
+          var hasUserPassedTest = function() {
+            var isTestPassed = true;
+            if (isInPassOrFailMode()) {
+              Object.keys(ctrl.scorePerSkillMapping).forEach(function(skillId) {
+                var correctionRate = ctrl.scorePerSkillMapping[skillId].score /
+                  ctrl.scorePerSkillMapping[skillId].total;
+                if (correctionRate <
+                  ctrl.questionPlayerConfig.questionPlayerMode.passCutoff) {
+                  isTestPassed = false;
+                }
+              });
+            }
+
+            if (!isTestPassed) {
+              ctrl.questionPlayerConfig.resultActionButtons = [];
+            }
+            return isTestPassed;
+          };
+
+          ctrl.getScorePercentage = function(scorePerSkill) {
+            return scorePerSkill.score / scorePerSkill.total * 100;
+          };
+
+          ctrl.getColorForScore = function(scorePerSkill) {
+            var correctionRate = scorePerSkill.score / scorePerSkill.total;
+            if (correctionRate >=
```

Comment:
Okay so suppose questionPlayerMode doesn't exist, this function will fail.
To account for that, always return the green color if questionPlayerMode doesn't exist. Then do all this logic in case questionPlayerMode does exist.

Reply:
Done.